Reduce complexity of grammar

stefanrueger opened the pull request:
This PR shifts simple programmer, part or memory components from `config_gram.y` to a table in `config.c`. Both methods have been used in parallel for a while. The latter method is less complex and easier to maintain.

This PR also prints warning messages on outdated avrdude.conf grammar use, eg, `has_debugwire = yes;` for a part will now print a warning. 

Reviewing this PR is easy: use the developer options to create avrdude's understanding of the config file with and without this PR, then compare the outputs:
```
$ avrdude_git -p*/s -c*/s >/tmp/avrdude.conf.git
$ avrdude_pr  -p*/s -c*/s >/tmp/avrdude.conf.pr
$ diff /tmp/avrdude.conf.git /tmp/avrdude.conf.pr
```

mcuee commented:
Tested under MSYS2 mingw and it is good.

```
$ diff -u avrdude.conf.git avrdude.pr1282.git
$ ./avrdude_git -p*/s -c*/s >/tmp/avrdude.conf.git
$ ./avrdude_pr1282 -p*/s -c*/s >/tmp/avrdude.pr1282.git
$ diff -u /tmp/avrdude.conf.git /tmp/avrdude.pr1282.git

$ ./avrdude_pr1282 -C avrdude7.conf -p*/s -c*/s >test.txt
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:404]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:405]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:433]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:434]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:466]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:467]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:488]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:489]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:552]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:553]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:572]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:573]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:591]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:592]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:622]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:623]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:676]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:677]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:760]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:762]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:771]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:773]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:783]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:785]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:796]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:798]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:808]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:809]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:831]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:833]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:855]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:857]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1436]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1437]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1458]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1459]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1472]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1473]
avrdude_pr1282 warning: mosi is deprecated, will be removed in v8.0, use sdo [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1486]
avrdude_pr1282 warning: miso is deprecated, will be removed in v8.0, use sdi [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1487]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:1711]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:3202]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:3391]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:3580]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:3772]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:3964]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:4157]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:4343]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:4667]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:4886]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:5086]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:5442]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:5654]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:5908]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:6142]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:7107]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:7304]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:7501]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:7758]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:7985]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:8212]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:8441]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:8638]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:8824]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:9011]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:9199]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:9387]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:9728]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:9926]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:10110]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:10393]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:10577]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:10760]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:10946]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:11136]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:11339]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:11781]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:11974]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:12167]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:12483]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:12666]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:12857]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13048]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13252]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13454]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13455]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13641]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13642]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13830]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:13831]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14018]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14019]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14206]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14207]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14390]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14612]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:14832]
avrdude_pr1282 warning: has_pdi is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15111]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15239]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15328]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15428]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15588]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15646]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15758]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15890]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:15990]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16267]
avrdude_pr1282 warning: is_avr32 is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16268]
avrdude_pr1282 warning: has_debugwire is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16292]
avrdude_pr1282 warning: has_tpi is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16490]
avrdude_pr1282 warning: has_jtag is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16647]
avrdude_pr1282 warning: has_updi is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:16712]
avrdude_pr1282 warning: has_updi is deprecated, will be removed in v8.0, use prog_modes [C:\work\avr\avrdude_test\avrdude_bin\avrdude7.conf:18004]


```

MCUdude commented:
I haven't tried the PR, but does this introduce breaking changes if an older avrdude.conf file is used from e.g Avrdude 6.3 or 7.0?

stefanrueger commented:
No. It shouldn't!

I have added courtesy warnings, though, that certain grammar constructs are not going to be supported from v8.0.

Older avrdude.conf files should still be parsed and accepted(*).

Generally, everyone should realise that by using older conf files
 - Certain features will be unavailable (eg, hvupdi_support)
 - Bugs that came from old avrdude.conf files will still be there
 - Bugs that were addressed by adding avrdude.conf parameters will resurface

I never understood the utility of insisting on using old .conf files with new code binaries. I realise sometimes people have made private changes to avrdude.conf entries to trick AVRDUDE to do this or that (add ISP commands to UPDI parts, set certain delays to seconds etc), but that's not a great reason to dampen forward development of avrdude.conf.

(*) There is one exception: At some point in the history of AVRDUDE the word bank has been replaced with page in avrdude.conf (banked -> paged, num_banks -> num_pages, bank_size -> page_size). I have removed backward compatibility of the bank aliases b/c I didn't find a single mention in the documentation about that.


On Sat, Jan 14, 2023 at 18:04, Hans wrote:
| Date: Sat, 14 Jan 2023 10:04:10 -0800
| From: Hans ***@***.***>
| To: avrdudes/avrdude ***@***.***>
| Cc: Stefan Rueger ***@***.***>, State change
|  ***@***.***>
| Subject: Re: [avrdudes/avrdude] Reduce complexity of grammar (PR #1282)
| 
| I haven't tried the PR, but does this introduce breaking changes if an older avrdude.conf file is used from e.g Avrdude 6.3 or 7.0?
| 
| -- 
| Reply to this email directly or view it on GitHub:
| https://github.com/avrdudes/avrdude/pull/1282#issuecomment-1382873645
| You are receiving this because you modified the open/close state.
| 
| Message ID: ***@***.***>
